cache/src/test/kotlin/com/dropbox/android/external/cache4/CacheLoaderTest.kt

```diff
@@ -0,0 +1,314 @@
+package com.dropbox.android.external.cache4
+
+import com.google.common.truth.Truth.assertThat
+import com.nhaarman.mockitokotlin2.never
+import com.nhaarman.mockitokotlin2.spy
+import com.nhaarman.mockitokotlin2.times
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.whenever
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.junit.Assert.assertThrows
+import org.junit.Test
+import java.io.IOException
+import java.util.concurrent.TimeUnit
+
+class CacheLoaderTest {
+
+    private val clock = TestClock(virtualTimeNanos = 0)
+    private val expiryDuration = TimeUnit.MINUTES.toNanos(1)
+
+    @Test
+    fun `get(key, loader) returns value from loader when no entry with the associated key exists before and after executing the loader`() {
+        val cache = CacheBuilder()
+            .build<Long, String>()
+
+        val loader = createLoader("dog")
+
+        val value = cache.get(1, loader)
+
+        verify(loader, times(1)).invoke()
+
+        assertThat(value)
+            .isEqualTo("dog")
+    }
+
+    @Test
+    fun `get(key, loader) returns existing value when an entry with the associated key exists before executing the loader`() {
+        val cache = CacheBuilder()
+            .build<Long, String>()
+
+        cache.put(1, "dog")
+
+        val loader = createLoader("cat")
+
+        val value = cache.get(1, loader)
+
+        verify(loader, never()).invoke()
+
+        assertThat(value)
+            .isEqualTo("dog")
+    }
+
+    @Test
+    fun `get(key, loader) returns existing value when an entry with the associated key is absent initially but present after executing the loader`() =
+        runBlocking {
```

Comment:
I wanted to use `TestCoroutineScope.runBlockingTest` here to advance the delays, but it doesn't seem to wait for the coroutines to finish when I'm launching with a different dispatcher: `launch(Dispatchers.Default)` which I need to be able to test concurrency. Is there a better way to test this?

Reply:
@yigit has strategies for testing async coroutines maybe he can chime in